### PR TITLE
M-FAC documentation and Tutorials

### PR DESCRIPTION
**Pull request by bfineran**
* README doc that moves over most of the information from #282 
* Gradual pruning tutorial with MobileNet Imagenette that shows a ~2% Top-1 Acc difference on an short, aggressive pruning task
* One-Shot pruning tutorial on MNISTNet that shows M-FAC dropping >2% accuracy while magnitude drops <50% accuracy when pruning to 35% sparsity

### CODE REVIEW

File: research/mfac/README.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+software distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+# Matrix-Free Approximate Curvature (M-FAC) Pruning
+
+The paper
+[Efficient Matrix-Free Approximations of Second-Order Information, With Applications to Pruning and Optimization](https://arxiv.org/pdf/2107.03356.pdf)
```
Comment (jeanniefinks) [inline research/mfac/README.md:20]: Since "With" can be capped or not depending on which writing style is used, looking at the paper, they've lowercased it to be "with" I think.

File: research/mfac/README.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+This algorithm is shown to outperform magnitude pruning as well as other second-order pruning
+techniques on a variety of one-shot and gradual pruning tasks.
+
+## Using M-FAC with SparseML
+
+SparseML makes it easy to use the M-FAC pruning algorithm as part of sparsification
+recipes to improve pruning recovery by providing an `MFACPruningModifier`.
+The `MFACPruningModifier` contains the same settings as the magnitude
+pruning modifiers and contains extra settings for the M-FAC algorithm under the
+`mfac_options` parameter.  `mfac_optons` should be provided as a YAML dictionary and
```
Comment (jeanniefinks) [inline research/mfac/README.md:35]:
```suggestion
`mfac_options` parameter.  `mfac_options` should be provided as a YAML dictionary and
```

File: research/mfac/README.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+details of the main options will be provided below.
```
Comment (jeanniefinks) [inline research/mfac/README.md:36]:
```suggestion
details of the main options are provided below.
```

File: research/mfac/README.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+available to run on and model to be pruned.
+
+#### num_grads
+To approximate the second order information in the M-FAC algorithm, first order
+gradients are used. `num_grads` specifies the number of recent gradient samples to store
+of a model while training.
+
+This value can be an int where that constant value will be used throughout pruning or a
+dictionary of float sparsity values to the number of gradients that should be
+stored when that sparsity level (between 0.0 and 1.0) is reached. If a
```
Comment (jeanniefinks) [inline research/mfac/README.md:69]: consider breaking up this intro sentence; it sounds a little winded 😺

File: research/mfac/recipes/pruning-mobilenet-imagenette-mfac-short-95.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+---
+
+# Pruning MobileNet-Imagenette M-FAC
+This recipe prunes a MobileNet model to 95% sparsity over 3 epochs using the M-FAC algorithm.
+
+* This recipe must be run at train batch size 32. If running at a different batch size,
+    the learning rate and number of M-FAC gradients should be adjusted accordingly
+* `available_gpus` should be updated based on the devices available on the system.
+    If no GPU is available, then it should be removed
+* This recipe is for demonstration purposes only, in practice a larger dataset and longer pruning
```
Comment (jeanniefinks) [inline research/mfac/recipes/pruning-mobilenet-imagenette-mfac-short-95.md:70]:
```suggestion
* This recipe is for demonstration purposes only; in practice a larger dataset and longer pruning
```

File: research/mfac/tutorials/gradual_pruning_with_mfac.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+high sparsities with minimal accuracy degradation.
+Instead of magnitudes, the M-FAC pruning algorithm efficiently uses first-order model
+information (gradients) to approximate second-order information that can be used to
+decide which weights are optimal to prune. This technique can often lead to superior
+results to magnitude pruning when pruning to higher sparsities or when pruning with fewer
+epochs.
+
+In this tutorial, you will prune a MobileNet model trained on the tiny Imagenette dataset - 
+once with magnitude pruning and once with M-FAC pruning.  Both pruning runs will
+prune the model to 95% sparsity and only use 3 epochs . The selected model, dataset, and
```
Comment (jeanniefinks) [inline research/mfac/tutorials/gradual_pruning_with_mfac.md:35]:
```suggestion
prune the model to 95% sparsity and only use 3 epochs. The selected model, dataset, and
```

File: research/mfac/tutorials/gradual_pruning_with_mfac.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+pruning schedule are selected for the purposes of this tutorial only to quickly demonstrate
+the benefits of the M-FAC algorithm.  In practice, a longer pruning schedule and larger dataset
+should be used.
+
+
+### Steps
+1. Setting Up
+2. Inspecting the Recipes
+3. Run M-FAC and Magnitude Pruning
```
Comment (jeanniefinks) [inline research/mfac/tutorials/gradual_pruning_with_mfac.md:44]:
```suggestion
3. Running M-FAC and Magnitude Pruning
```

File: research/mfac/tutorials/gradual_pruning_with_mfac.md [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+in SparseML. They can be viewed with the command line or by following the links below:
+
+* M-FAC: [sparseml/research/mfac/recipes/pruning-mobilenet-imagenette-mfac-short-95.md](https://github.com/neuralmagic/sparseml/blob/main/research/mfac/recipes/pruning-mobilenet-imagenette-mfac-short-95.md)
+* Magnitude: [sparseml/research/mfac/recipes/pruning-mobilenet-imagenette-magnitude-short-95.md](https://github.com/neuralmagic/sparseml/blob/main/research/mfac/recipes/pruning-mobilenet-imagenette-magnitude-short-95.md)
+
+Notice that the recipes are identical except that the `MFACPruningModifier` and `mfac_options`
+are used in M-FAC pruning.  The M-FAC recipe's `available_gpus` parameter should be updated
+based on your system's available devices. If no GPU is available, then the line may be deleted.
+
+## Run M-FAC and Magnitude Pruning
```
Comment (jeanniefinks) [inline research/mfac/tutorials/gradual_pruning_with_mfac.md:75]:
```suggestion
## Running M-FAC and Magnitude Pruning
```

File: research/mfac/tutorials/one_shot_pruning_with_mfac.md [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+For this tutorial, the recipes for pruning MNISTNet to 35% sparsity are provided
+in SparseML. They can be viewed with the command line or by following the links below:
+
+* M-FAC: [research/mfac/recipes/pruning-mnistnet-one_shot-mfac.md](https://github.com/neuralmagic/sparseml/blob/main/research/mfac/recipes/pruning-mnistnet-one_shot-mfac.md)
+* Magnitude: [research/mfac/recipes/pruning-mnistnet-one_shot-magnitude.md](https://github.com/neuralmagic/sparseml/blob/main/research/mfac/recipes/pruning-mnistnet-one_shot-magnitude.md)
+
+Notice that the recipes are identical except that the `MFACPruningModifier` and `mfac_options`
+are used in M-FAC pruning.
+
+The remainder of this tutorial can be run in a single python REPL
```
Comment (jeanniefinks) [inline research/mfac/tutorials/one_shot_pruning_with_mfac.md:72]:
```suggestion
The remainder of this tutorial can be run in a single Python REPL
```

File: research/mfac/tutorials/one_shot_pruning_with_mfac.md [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+
+```python
+from sparseml.pytorch.optim import ScheduledModifierManager
+
+manager_mfac = ScheduledModifierManager.from_yaml("research/mfac/recipes/pruning-mnistnet-one_shot-mfac.md")
+manager_mfac.apply(model_mfac, grad_sampler=grad_sampler)
+```
+
+## One-Shot Pruning with Magnitude Pruning
+Magnitude pruning recipes just look at the pretrained model weights, so need
```
Comment (jeanniefinks) [inline research/mfac/tutorials/one_shot_pruning_with_mfac.md:144]:
```suggestion
Magnitude pruning recipes just look at the pre-trained model weights, so need
```

File: research/mfac/tutorials/one_shot_pruning_with_mfac.md [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+no extra information to prune the model in one-shot. The recipe can be applied directly
```
Comment (jeanniefinks) [inline research/mfac/tutorials/one_shot_pruning_with_mfac.md:145]:
```suggestion
information is not needed to prune the model in one-shot. The recipe can be applied directly
```

File: research/mfac/tutorials/one_shot_pruning_with_mfac.md [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+* Magnitude: [research/mfac/recipes/pruning-mnistnet-one_shot-magnitude.md](https://github.com/neuralmagic/sparseml/blob/main/research/mfac/recipes/pruning-mnistnet-one_shot-magnitude.md)
+
+Notice that the recipes are identical except that the `MFACPruningModifier` and `mfac_options`
+are used in M-FAC pruning.
+
+The remainder of this tutorial can be run in a single python REPL
+(launch by entering `python` in your command prompt).
+
+## Applying Recipes in One Shot
+All SparseML recipes can be applied to a model in one-shot by using
```
Comment (jeanniefinks) [inline research/mfac/tutorials/one_shot_pruning_with_mfac.md:76]:
```suggestion
All SparseML recipes can be applied to a model in one shot by using
```

File: research/mfac/README.md (hunk 1 above)
Comment (bfineran) [inline research/mfac/README.md:20]:
```suggestion
[Efficient Matrix-Free Approximations of Second-Order Information, with Applications to Pruning and Optimization](https://arxiv.org/pdf/2107.03356.pdf)
```

File: research/mfac/README.md [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+dictionary, then 0.0 must be included as a key for the base number of gradients
+to store (i.e. {0: 64, 0.5: 128, 0.75: 256}).
```
Comment (bfineran) [inline research/mfac/README.md:71]:
```suggestion
This value can be an int where that constant value will be used throughout pruning.
Alternatively the value can be a dictionary of float sparsity values to the number of
gradients that should be stored when that sparsity level (between 0.0 and 1.0) is reached.
If a dictionary is used, then 0.0 must be included as a key for the base number of gradients
to store (i.e. {0: 64, 0.5: 128, 0.75: 256}).
```

**Review by jeanniefinks — APPROVED** [review]

**Review by markurtz — APPROVED** [review]
This looks great Ben!
